### PR TITLE
feat: adding a dynamic sleep interval in the connectivity loop

### LINKED ISSUE #2877: chore: dynamically increase ConnectivityLoopInterval value
### Background
We try to connect to peers every fixed amount of time, which is defined by the following const
https://github.com/waku-org/nwaku/blob/49f03f124349250fb29e3b642a0d3f84e4045fc6/waku/node/peer_manager/peer_manager.nim#L62

The more peers we have, the less we need to run this loop. The inverse is also true - the less peers we have, the more we need to run this.

We should dynamically modify this waiting time depending on how many peers we have

### Acceptance criteria
- [ ] the connectivity loop wait time is dynamic based on current amount of peers

### PR DESCRIPTION
**Pull request by gabrielmer**
# Description

Dynamically setting the sleep time for the next run of the procs that manage peer connections and disconnections.
The farther away we are from our targets, the less time it will pass until the next attempt to connect to peers.

If we're already at our target, we will sleep the whole `ConnectivityLoopInterval`

# Changes

- [x] decreasing the interval between connectivity loop iterations as a function of remaining `out` connections to reach our target
- [x] decreasing the interval between connectivity loop iterations as a function of excess `in` connections

## Issue

closes #2877

### CODE REVIEW

File: waku/node/peer_manager/peer_manager.nim [hunk 1]
```diff
@@ -965,7 +965,21 @@ proc relayConnectivityLoop*(pm: PeerManager) {.async.} =
...
+      excessInConns = max(inRelayPeers.len - pm.inRelayPeersTarget, 0)
+
+      # One minus the percentage of excess connections relative to the target, limited to 100% 
+      # We calculate one minus this percentage because we want the factor to be inversely proportional to the number of excess peers
+      inFactor = 1 - min(excessInConns / pm.inRelayPeersTarget, 1)
+      # Percentage of out relay peers relative to the target
+      outFactor = min(outRelayPeers.len / pm.outRelayPeersTarget, 1)
+      factor = min(outFactor, inFactor)
+      dynamicSleepInterval =
+        chronos.seconds(int(float(ConnectivityLoopInterval.seconds()) * factor))
```
Comment (gabrielmer) [inline waku/node/peer_manager/peer_manager.nim:979]: Not a big fan of these casts, couldn't think of a simpler and nicer way, but open for suggestions :))

**Comment by gabrielmer** [comment]
@SionoiS suggested in https://github.com/waku-org/nwaku/issues/2877#issuecomment-2213795222 to also trigger the connectivity loop if a node has many peers and just added many other ones recently. Or in other words, make the interval of a function both of the number of peers and of how recently those peers were added.

Thinking it through and looking at the implementation, I think that adds complexity and don't think that there's actual performance benefits to it, I think that what matters is the amount of active connections and not really when they were established. But open for opinions/ideas/corrections :)

**Review by Ivansete-status — COMMENTED** [review]
Thanks so much for it! :raised_hands: 
Nevertheless, I wonder if we could make it a bit simpler. I was envisioning just a delay factor depending on how far we are from the max target ( in + out ) and, if we are currently beyond the target then keep the fixed big delay.
Cheers

EDITED: actually, the farther we are from the number connection target (in+out) the faster we need to strive to reach the target again :) It is great how you are considering the exceed num

**Comment by SionoiS** [comment]
> @SionoiS suggested in [#2877 (comment)](https://github.com/waku-org/nwaku/issues/2877#issuecomment-2213795222) to also trigger the connectivity loop if a node has many peers and just added many other ones recently. Or in other words, make the interval of a function both of the number of peers and of how recently those peers were added.

The benefit is balancing the # of peers per shards, since we can't know their shard before connecting via Waku Metadata but it's true that is not strictly needed.

**Comment by gabrielmer** [comment]
There was a proposal 

> Thanks so much for it! 🙌 Nevertheless, I wonder if we could make it a bit simpler. I was envisioning just a delay factor depending on how far we are from the max target ( in + out ) and, if we are currently beyond the target then keep the fixed big delay. Cheers
> 
> EDITED: actually, the farther we are from the number connection target (in+out) the faster we need to strive to reach the target again :) It is great how you are considering the exceed num

Yes! So it ended up looking more complicated than what I originally expected, I added the comments so it's clearer because I couldn't think of a way of doing the same with simpler math.

What I did was basically multiply a factor between 0 and 1 that depends on how far away we are from our targets and reduces the wait the farther away we are from the targets. In theory it's simple, but the excess peers case made it a bit complicated, as it should be inversely proportional to the factor, and at the same time excess peers are not bounded (it can be over 100% of the target). These two properties added extra operations that aren't that intuitive.

My first attempt was with the sum of `in` + `out`, but because both have different targets and behave differently (in one we check that we don't have too few and in the other that we don't have too many), I got the idea that it's like adding apples to oranges and that it should be calculated separately and take the minimum factor out of both of them.

Can you please be a bit more specific on how to simplify the math? I couldn't find a simpler way myself with the math working out, but would love to hear a better way :))

**Comment by gabrielmer** [comment]
> The benefit is balancing the # of peers per shards, since we can't know their shard before connecting via Waku Metadata but it's true that is not strictly needed.

Aren't the shards in the ENR? I thought that balancing the connections by shards is what's done by https://github.com/waku-org/nwaku/blob/1713f562353476235a2076941775a1f150d7c615/waku/node/peer_manager/peer_manager.nim#L763-L839

Maybe I'm mixing things up

**Review by Ivansete-status — APPROVED** [review]

**Comment by SionoiS** [comment]
> Aren't the shards in the ENR? I thought that balancing the connections by shards is what's done by

Yes it true but Waku Metadata might not reflect the ENR shard advertised but I guess it's not a big deal.

**Pull request merged**